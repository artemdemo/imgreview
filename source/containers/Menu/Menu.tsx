import React from 'react';
import {connect} from 'react-redux';
import {TReduxState} from '../../reducers';
import MIOpenImage from '../MenuItems/MIOpenImage.async';
import MISave from '../MenuItems/MISave';
import MIArrow from '../MenuItems/MIArrow.async';
import MIStrokeColor from '../MenuItems/MIStrokeColor';
import MIStrokeWidth from '../MenuItems/MIStrokeWidth';
import MIResize from '../MenuItems/MIResize/MIResize';
import MIGithub from '../MenuItems/MIGithub';
import MainMenu from '../../components/MainMenu/MainMenu';
import {blurShapes, TBlurShapes} from '../../model/shapes/shapesActions';
import {TStateCanvas} from '../../model/canvas/canvasReducer';

type Props = {
    canvas: TStateCanvas;
    blurShapes: TBlurShapes;
};

class Menu extends React.PureComponent<Props> {
    onMenuClick = () => {
        const { blurShapes } = this.props;
        blurShapes();
    };

    render() {
        const { canvas } = this.props;
        const disabled = canvas.image == null;
        return (
            <MainMenu
                onClick={this.onMenuClick}
            >
                <MIOpenImage />
                <MISave disabled={disabled} />
                <MIArrow disabled={disabled} />
                <MIStrokeColor disabled={disabled} />
                <MIStrokeWidth disabled={disabled} />
                <MIResize disabled={disabled} />
                <MIGithub />
            </MainMenu>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        canvas: state.canvas,
    }), {
        blurShapes,
    },
)(Menu);
