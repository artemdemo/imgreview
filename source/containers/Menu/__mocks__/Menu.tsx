import React from 'react';

const Menu = props => (
    <div data-mock='Menu'>
        {props.children}
    </div>
);

export default Menu;
