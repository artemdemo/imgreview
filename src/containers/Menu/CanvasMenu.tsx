import React, { useContext, useState } from 'react';
import { CanvasMenuEvents } from './CanvasMenuEvents';
import { AppStateContext } from '../../model/AppStateContext';
import IShape from '../../../srcCanvas/canvasShapes/Shape/IShape';
import { setShapeToAdd } from '../../model/menu/menuActions';
import { EShapeTypes } from '../../../srcCanvas/api/api-types';
import { TopMenuGroup } from '../../components/TopMenu/TopMenuGroup';
import { MIOpenImage } from '../MenuItems/MIOpenImage';
import { MISave } from '../MenuItems/MISave/MISave';
import { MICopyAll } from '../MenuItems/MICopyAll';
import { MIArrow } from '../MenuItems/MIArrow';
import { MIText } from '../MenuItems/MIText';
import { MIRect } from '../MenuItems/MIRect';
import { MIEllipse } from '../MenuItems/MIEllipse';
import { MIStrokeColor } from '../MenuItems/MIStrokeColor/MIStrokeColor';
import { MIStrokeWidth } from '../MenuItems/MIStrokeWidth';
import { MIFontSize } from '../MenuItems/MIFontSize';
import { MISketchify } from '../MenuItems/MISketchify';
import { MIBringFront } from '../MenuItems/MIBringFront';
import { MISendBack } from '../MenuItems/MISendBack';
import { MIBlankCanvas } from '../MenuItems/MIBlankCanvas';
import { useDevQueries } from '../../services/url';
import { MIFillColor } from '../MenuItems/MIFillColor';

type State = {
  showStrokeColor: boolean;
  showStrokeWidth: boolean;
  showFillColor: boolean;
  showCrop: boolean;
  showFontSize: boolean;
  showSketchify: boolean;
  showOrderButtons: boolean;
  clickedShapeType: EShapeTypes | undefined;
};

const initState: State = {
  showStrokeColor: false,
  showStrokeWidth: false,
  showFillColor: false,
  showCrop: false,
  showFontSize: false,
  showSketchify: false,
  showOrderButtons: false,
  clickedShapeType: undefined,
};

export const CanvasMenu: React.FC = () => {
  const [menuState, setMenuState] = useState<State>({ ...initState });
  const [hasShapes, setHasShapes] = useState<boolean>(false);
  const {
    state: {
      canvas: { canvasApi },
    },
    dispatch,
  } = useContext(AppStateContext);
  const { isDev } = useDevQueries();

  const setItemsVisibility = (shape?: IShape) => {
    dispatch(setShapeToAdd());

    const newState = {
      ...initState,
      showOrderButtons: false,
      clickedShapeType: shape?.type,
    };
    switch (shape?.type) {
      case EShapeTypes.ARROW:
      case EShapeTypes.RECT_ROUGH:
      case EShapeTypes.ELLIPSE_ROUGH:
        newState.showStrokeColor = true;
        newState.showStrokeWidth = true;
        if (shape?.type !== EShapeTypes.ARROW) {
          newState.showFillColor = true;
        }
        break;
      case EShapeTypes.RECT:
      case EShapeTypes.ELLIPSE:
        newState.showStrokeColor = true;
        newState.showStrokeWidth = true;
        newState.showFillColor = true;
        newState.showSketchify = true;
        break;
      case EShapeTypes.TEXT:
        newState.showStrokeColor = true;
        newState.showFontSize = true;
        break;
      case EShapeTypes.SELECT_RECT:
        newState.showCrop = true;
        break;
    }
    if (!!shape) {
      newState.showOrderButtons = true;
    }
    setMenuState(newState);
  };

  return (
    <>
      <CanvasMenuEvents
        onShapesAmountChanged={async () => {
          if (canvasApi) {
            const shapesAmount = await canvasApi.getShapesAmount();
            setHasShapes(shapesAmount > 0);
          }
        }}
        onShapeFocus={setItemsVisibility}
      />
      <TopMenuGroup>
        <MIOpenImage />
        <MISave disabled={!hasShapes} />
        <MICopyAll disabled={!hasShapes} />
      </TopMenuGroup>
      <TopMenuGroup>
        <MIArrow />
        <MIText />
        <MIRect />
        <MIEllipse />
      </TopMenuGroup>
      <TopMenuGroup>
        {menuState.showStrokeColor && <MIStrokeColor />}
        {menuState.showFillColor && <MIFillColor />}
        {menuState.showStrokeWidth && <MIStrokeWidth />}
        {menuState.showFontSize && <MIFontSize />}
        {menuState.showSketchify && (
          <MISketchify
            reverse={
              menuState.clickedShapeType === EShapeTypes.RECT_ROUGH ||
              menuState.clickedShapeType === EShapeTypes.ELLIPSE_ROUGH
            }
          />
        )}
      </TopMenuGroup>
      {menuState.showOrderButtons && (
        <TopMenuGroup>
          <MIBringFront />
          <MISendBack />
        </TopMenuGroup>
      )}
      {isDev && <MIBlankCanvas />}
    </>
  );
};
