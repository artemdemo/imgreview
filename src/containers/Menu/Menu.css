.Menu {
  padding: 4px 15px;
  background-color: var(--main-menu-color);
  z-index: var(--main-menu-zindex);
  position: fixed;
  left: 0;
  right: 0;
}
.Menu::after {
  content: '';
  display: table;
  clear: both;
}

.Menu__RightGroup {
  float: right;
}
