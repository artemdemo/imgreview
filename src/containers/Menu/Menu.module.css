.Menu {
  padding: 4px 15px;
  background-color: var(--main-menu-color);
  z-index: var(--main-menu-zindex);
}
.Menu::after {
  content: '';
  display: table;
  clear: both;
}

.Menu__RightGroup {
  float: right;
}
