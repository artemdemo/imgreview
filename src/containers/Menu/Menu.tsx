import React from 'react';
import { connect } from 'react-redux';
import { TReduxState } from '../../reducers';
import MIOpenImage from '../MenuItems/MIOpenImage';
import MISave from '../MenuItems/MISave';
import MIArrow from '../MenuItems/MIArrow';
import MIText from '../MenuItems/MIText';
import MIStrokeColor from '../MenuItems/MIStrokeColor';
import MIStrokeWidth from '../MenuItems/MIStrokeWidth';
import MIResize from '../MenuItems/MIResize/MIResize';
import MIGithub from '../MenuItems/MIGithub';
import MIBlankCanvas from '../MenuItems/MIBlankCanvas';
import TopMenuPanel from '../../components/TopMenu/TopMenuPanel';
import FloatRight from '../../components/Floating/FloatRight';
import { TStateCanvas } from '../../model/canvas/canvasReducer';
import { setMenuHeight, TSetMenuHeight } from '../../model/menu/menuActions';
import * as shapesService from '../../services/shapes'
import { isDev } from '../../services/env';
import * as canvasApi from '../../../srcCanvas/api';

type Props = {
    canvas: TStateCanvas;
    setMenuHeight: TSetMenuHeight
};

class Menu extends React.PureComponent<Props> {
    #menuRef = React.createRef<HTMLDivElement>();

    componentDidMount(): void {
        const { setMenuHeight } = this.props;
        const { offsetHeight } = this.#menuRef.current || {};
        if (offsetHeight) {
            setMenuHeight(offsetHeight);
        }

        // @ts-ignore
        canvasApi.shapesBlurred.on(() => {
            console.log('shapesBlurred');
        });

        // @ts-ignore
        canvasApi.shapeClicked.on((shape) => {
            console.log('shapeClicked', shape);
        });
    }

    onMenuClick = () => {
        shapesService.blurShapes();
    };

    renderBlankCanvas() {
        if (isDev) {
            return (
                <MIBlankCanvas />
            );
        }
        return null;
    }

    render() {
        const { canvas } = this.props;
        const disabled = canvas.height === 0 && canvas.width === 0;
        return (
            <TopMenuPanel
                onClick={this.onMenuClick}
                ref={this.#menuRef}
            >
                <MIOpenImage />
                <MISave disabled={disabled} />
                <MIArrow disabled={disabled} />
                <MIText disabled={disabled} />
                <MIStrokeColor disabled={disabled} />
                <MIStrokeWidth disabled={disabled} />
                <MIResize disabled={disabled} />
                {this.renderBlankCanvas()}
                <FloatRight>
                    <MIGithub />
                </FloatRight>
            </TopMenuPanel>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        canvas: state.canvas,
    }), {
        setMenuHeight,
    },
)(Menu);
