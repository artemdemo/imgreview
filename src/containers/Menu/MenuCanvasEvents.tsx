import React, { useContext, useEffect } from 'react';
import _ from 'lodash';
import { AppStateContext } from '../../model/AppStateContext';
import IShape from '../../../srcCanvas/canvasShapes/Shape/IShape';

type Props = {
  onShapesAmountChanged: () => void;
  onShapeFocus: (shape?: IShape) => void;
};

export const MenuCanvasEvents: React.FC<Props> = (props) => {
  const { onShapesAmountChanged, onShapeFocus } = props;
  const {
    state: {
      canvas: { canvasApi },
    },
  } = useContext(AppStateContext);

  useEffect(() => {
    let unsubShapesBlurred = _.noop;
    let unsubShapeClicked = _.noop;
    let unsubShapeDragStared = _.noop;
    let unsubShapeAdded = _.noop;
    let unsubShapeDeleted = _.noop;

    if (canvasApi) {
      unsubShapesBlurred = canvasApi.onShapesBlurred(onShapeFocus);
      unsubShapeClicked = canvasApi.onShapeClicked((shape) => {
        requestAnimationFrame(() => onShapeFocus(shape));
      });
      unsubShapeDragStared = canvasApi.onShapeDragStarted((shape) => {
        requestAnimationFrame(() => onShapeFocus(shape));
      });
      unsubShapeAdded = canvasApi.onShapeAdded(async ({ addedShape }) => {
        onShapeFocus(addedShape);
        onShapesAmountChanged();
      });
      unsubShapeDeleted = canvasApi.onShapeDeleted(onShapesAmountChanged);
      onShapesAmountChanged();
    }

    return () => {
      unsubShapesBlurred();
      unsubShapeClicked();
      unsubShapeDragStared();
      unsubShapeAdded();
      unsubShapeDeleted();
    };
  }, [canvasApi]);

  return null;
};
