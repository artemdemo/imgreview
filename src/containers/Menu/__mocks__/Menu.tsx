import React from 'react';

const Menu: React.FC = (props) => <div data-mock="Menu">{props.children}</div>;

export default Menu;
