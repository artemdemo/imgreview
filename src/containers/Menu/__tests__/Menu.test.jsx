import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Menu from '../Menu';

jest.mock('../../MenuItems/MIOpenImage');
jest.mock('../../MenuItems/MISave');
jest.mock('../../MenuItems/MIArrow');
jest.mock('../../MenuItems/MIText');
jest.mock('../../MenuItems/MIStrokeColor');
jest.mock('../../MenuItems/MIStrokeWidth');
jest.mock('../../MenuItems/MIResize/MIResize');
jest.mock('../../MenuItems/MIGithub');
jest.mock('../../MenuItems/MIBlankCanvas');
jest.mock('../../../components/TopMenu/TopMenuPanel');
jest.mock('../../../components/Floating/FloatRight');
jest.mock('../../../../srcCanvas/api');

const apiMock = require('../../../../srcCanvas/api');

describe('Menu', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should render', () => {
        const tree = renderer.create(
            <Menu
                canvas={{
                    imageHeight: 10,
                    imageWidth: 10,
                }}
            />
        );

        expect(tree).toMatchSnapshot();
    });

    it('should handle click on the panel', () => {
        const wrapper = mount(
            <Menu
                canvas={{
                    imageHeight: 10,
                    imageWidth: 10,
                }}
            />
        );
        const instance = wrapper.instance();
        instance.onMenuClick();
        expect(apiMock.blurShapes).toBeCalledWith();
    });
});
