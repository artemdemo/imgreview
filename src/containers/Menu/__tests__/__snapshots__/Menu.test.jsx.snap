// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render 1`] = `
<div
  data-mock="TopMenuPanel"
>
  <div
    data-mock="MIOpenImage"
  >
    {}
  </div>
  <div
    data-mock="MISave"
  >
    {
  "disabled": false
}
  </div>
  <div
    data-mock="Separator"
  >
    {}
  </div>
  <div
    data-mock="MIArrow"
  >
    {
  "disabled": false
}
  </div>
  <div
    data-mock="MIText"
  >
    {
  "disabled": false
}
  </div>
  <div
    data-mock="MISelect"
  >
    {
  "disabled": false
}
  </div>
  <div
    data-mock="MIRect"
  >
    {
  "disabled": false
}
  </div>
  <div
    data-mock="Separator"
  >
    {}
  </div>
  <div
    data-mock="MIStrokeColor"
  >
    {
  "disabled": false,
  "show": false
}
  </div>
  <div
    data-mock="MIStrokeWidth"
  >
    {
  "disabled": false,
  "show": false
}
  </div>
  <div
    data-mock="Separator"
  >
    {
  "show": false
}
  </div>
  <div
    data-mock="MIResize"
  />
  <div
    data-mock="MIBlankCanvas"
  >
    {
  "show": false
}
  </div>
  <div
    data-mock="FloatRight"
  >
    <div
      data-mock="MIGithub"
    >
      {}
    </div>
  </div>
</div>
`;
